reconcile/gitlab_permissions.py

```diff
@@ -1,21 +1,127 @@
+    ) -> None:
+        self.gl = gl
+        self.group_name = group_name
+        self.access_level_string = access
+        self.access_level = self.gl.get_access_level(access)
+        self.dry_run = dry_run
+
+    def run(self, repos: list[str]) -> None:
+        # get repos not owned by app-sre
+        non_app_sre_project_repos = {
+            repo
+            for repo in repos
+            if not self.gl.is_group_project_owner(
+                group_name=self.group_name, repo_url=repo
+            )
+        }
+
+        desired_state = {
+            project_repo_url: GroupSpec(self.group_name, self.access_level)
+            for project_repo_url in non_app_sre_project_repos
+        }
+        group_id, shared_projects = self.gl.get_group_id_and_shared_projects(
+            self.group_name
+        )
+        current_state = {
+            project_repo_url: GroupSpec(
+                shared_projects[project_repo_url]["group_name"],
+                shared_projects[project_repo_url]["group_access_level"],
+            )
+            for project_repo_url in shared_projects
+        }
+        self.reconcile(desired_state, current_state, group_id)
+
+    def can_share_project(self, project: Project) -> bool:
+        # check if user have access greater or equal access required
+        user = project.members_all.get(id=self.gl.user.id)
+        return user.access_level >= self.access_level
+
+    def reconcile(
+        self,
+        desired_state: dict[str, GroupSpec],
+        current_state: dict[str, GroupSpec],
+        group_id: int,
+    ) -> None:
+        # get the diff data
+        diff_data = diff_mappings(
+            current=current_state,
+            desired=desired_state,
+            equal=lambda current, desired: current.group_access_level
+            == desired.group_access_level,
+        )
+
+        for repo in diff_data.add.keys():
+            project = self.gl.get_project(repo)
+            if not self.can_share_project(project):
+                logging.error(
+                    "%s is not shared with %s as %s",
+                    repo,
+                    self.gl.user.username,
+                    self.access_level_string,
+                )
+                return None
+            logging.info([
+                f"share_group_{self.group_name}_as_{self.access_level_string}",
+                repo,
+            ])
+            if not self.dry_run:
+                self.gl.share_project_with_group(
+                    project=project, group_id=group_id, access_level=self.access_level
+                )
+
+        for repo in diff_data.change.keys():
+            project = self.gl.get_project(repo)
+            if not self.can_share_project(project):
+                logging.error(
+                    "%s is not shared with %s as %s",
+                    repo,
+                    self.gl.user.username,
+                    self.group_name,
+                )
+                return None
```

Comment:
this will terminate loop and no exceptions throw, the integration will always success